File: dataset_reduzido/avatar/Python/Code/atcoder_AGC037_E.py
N , K = map ( int , input ( ).split ( ) )
S = input ( )
aaaa = False
if K >= 15 :
    aaaa = True
elif 2 ** K >= N :
    aaaa = True
if aaaa :
    print ( min ( S ) * N )
    exit ( )
def get_last_dict ( s_str ) :
    U = s_str + s_str [ : : - 1 ]
    c = min ( s_str )
    p = U.index ( c )
    minindex = p
    p += 1
    while p <= N :
        if U [ p ] == c :
            if check_normal_dict ( U , minindex , p ) :
                minindex = p
        p += 1
    return U [ minindex : minindex + N ]
def check_normal_dict ( u , pointer1 , pointer2 ) :
    for i in range ( N ) :
        if u [ pointer1 + i ] > u [ pointer2 + i ] :
            return True
        elif u [ pointer1 + i ] < u [ pointer2 + i ] :
            return False
    return False
S = get_last_dict ( S )
if K == 1 :
    print ( S )
else :
    count = 0
    for c in S :
        if c == S [ 0 ] :
            count += 1
        else :
            break
    if count * ( 2 ** ( K - 1 ) ) >= N :
        print ( S [ 0 ] * N )
    else :
        S = S [ 0 ] * ( count * ( ( 2 ** ( K - 1 ) ) - 1 ) ) + S
        print ( S [ 0 : N ] )

File: output/Llama-2-100/avatar/Python/Java/atcoder_AGC037_E.java
import java.util.Scanner;

public class atcoder_AGC037_E {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n, k;
        n = scanner.nextInt();
        k = scanner.nextInt();
        String s = scanner.nextLine();
        boolean aaaa = false;
        if (k >= 15) {
            aaaa = true;
        } else if (2 << k >= n) {
            aaaa = true;
        }
        if (aaaa) {
            System.out.println(min(s) * n);
            return;
        }
        String lastDict = getLastDict(s);
        if (k == 1) {
            System.out.println(lastDict);
        } else {
            int count = 0;
            for (char c : lastDict) {
                if (c == lastDict.charAt(0)) {
                    count++;
                } else {
                    break;
                }
            }
            if (count * (2 << (k - 1)) >= n) {
                System.out.println(lastDict.charAt(0) * n);
            } else {
                lastDict = lastDict.substring(0, n);
                System.out.println(lastDict);
            }
        }
    }

    public static String getLastDict(String s) {
        String u = s + s.substring(0, s.length() - 1);
        int minindex = 0;
        for (int i = 1; i <= s.length(); i++) {
            if (u.charAt(i) == s.charAt(minindex)) {
                minindex = i;
            }
        }
        return u.substring(minindex, minindex + s.length());
    }

    public static boolean checkNormalDict(String u, int pointer1, int pointer2) {
        for (int i = 0; i < s.length(); i++) {
            if (u.charAt(pointer1 + i) > u.charAt(pointer2 + i)) {
                return true;
            } else if (u.charAt(pointer1 + i) < u.charAt(pointer2 + i)) {
                return false;
            }
        }
        return false;
    }
}
